spectec/src/el/free.ml

```diff
@@ -115,7 +114,7 @@ and free_exp e =
     union (free_list free_exp [e1; e2]) (free_path p)
   | StrE efs -> free_nl_list free_expfield efs
   | CallE (id, args) -> union (free_defid id) (free_list free_arg args)
-  | IterE (e1, iter) -> union (free_exp e1) (free_iter iter)
+  | IterE (e1, iter) -> union (diff (free_exp e1) (binding_iter iter)) (free_iter iter)
```

Comment:
The semantics I had in mind so far was one where
```
(...)^(i<n)
```
effectively is a shorthand for
```
(...)^n /\ i^n = 0 ... (n-1)
```
i.e., `i` is simply a sequence of length n, with nothing else special about it. The advantage is that this avoids the introduction of local binders, which we do not currently have, and which might require more thought.

Was there a particular reason you had to change this interpretation? If not, it would probably make sense to split this change from this PR, so that we can discuss it separately.

Reply:
The fundamental reason was to fix a false type error.

For example, the following `sample.watsup` file
```
relation R: nat* -> nat*

var n: nat
var m: nat

rule R/r: n* -> m^8
  -- (if m = n*[i])^(i<8)
```
results in an unwanted type error:
```
$ ./watsup sample.watsup
sample.watsup:6.1-7.26: type error: rule contains unbound variable(s) `i`
```

which prevented writing reduction rules for one of SIMD instructions. (Specificaly, `VLOAD`)

After seeing this message, my first impression was: "the variable `i` should not be considered free in the entire `IterE` expression," and that's the reason for changing the implementation in a way that `i` is excluded from free variables in the entire `IterE` expression.

Since that was the only reason, I'm happy with any interpretation, as long as the example above is considered well-typed.

Reply:
Ah, okay. I'll think about it a little and tweak main to make that example work.

Reply:
This actually was intended to work, but there was a bug in the computation of bound variables, see #54.